wechaty-puppet-hostie/puppet_hostie.go

```diff
@@ -0,0 +1,965 @@
+  endpoint := p.option.Endpoint
+  if endpoint == "" {
+    ip, err := p.discoverHostieIP()
+    if err != nil {
+      return err
+    }
+    if ip == "" || ip == "0.0.0.0" {
+      return ErrNoEndpoint
+    }
+    endpoint = ip + ":8788"
+  }
+  conn, err := grpc.Dial(endpoint, grpc.WithInsecure())
+  if err != nil {
+    return err
+  }
+  p.grpcConn = conn
+  p.grpcClient = pbwechaty.NewPuppetClient(conn)
+  return nil
+}
+
+func (p *PuppetHostie) discoverHostieIP() (s string, err error) {
+  defer func() {
+    if err != nil {
+      err = fmt.Errorf("discoverHostieIP() err:%w", err)
+    }
+  }()
+  if p.option.Token == "" {
+    return "", errors.New("wechaty-puppet-hostie: token not found. See: <https://github.com/wechaty/wechaty-puppet-hostie#1-wechaty_puppet_hostie_token>")
+  }
+  const chatieEndpoint = "wss://api.chatie.io/v0/websocket/token/%s"
+  const protocol = "puppet-hostie|0.0.1"
+
+  dialer := websocket.Dialer{}
+  dialer.Subprotocols = append(dialer.Subprotocols, protocol)
+  conn, _, err := dialer.Dial(fmt.Sprintf(chatieEndpoint, p.option.Token), nil)
+  if err != nil {
+    return "", err
+  }
+  defer conn.Close()
+  err = conn.WriteJSON(map[string]interface{}{
+    "name": "hostie",
+  })
+  if err != nil {
+    return "", err
+  }
+
+  result := make(chan string)
+  errChan := make(chan error)
+  go func() {
+    for {
+      _, message, err := conn.ReadMessage()
+      if err != nil {
+        errChan <- err
+      }
+      resultMap := make(map[string]string)
+      err = json.Unmarshal(message, &resultMap)
+      if err != nil {
+        errChan <- err
+      }
+      if resultMap["name"] == "hostie" {
+        result <- resultMap["payload"]
+        return
+      }
+    }
+  }()
+
+  select {
+  case ip := <-result:
+    return ip, nil
+  case err := <-errChan:
+    return "", err
+  }
+}
+
+func (p *PuppetHostie) startGrpcStream() (err error) {
+  defer func() {
+    if err != nil {
+      err = fmt.Errorf("startGrpcStream err:%w", err)
+    }
+  }()
+  if p.eventStream != nil {
+    return errors.New("event stream exists")
+  }
+  p.eventStream, err = p.grpcClient.Event(context.Background(), &pbwechaty.EventRequest{})
+  if err != nil {
+    return err
+  }
+
+  go func() {
+    for {
+      reply, err := p.eventStream.Recv()
+      if err == io.EOF {
+        log.Println("eventStream.Recv EOF")
+        break
+      }
+      if err != nil {
+        log.Printf("PuppetHostie startGrpcStream() eventStream err %s", err)
+        reason := "startGrpcStream() eventStream err: " + err.Error()
+        p.option.Emit(schemas.PuppetEventNameReset, schemas.EventResetPayload{Data: reason})
+        break
+      }
+      go p.onGrpcStreamEvent(reply)
+    }
+  }()
+  return nil
+}
+
+func (p *PuppetHostie) onGrpcStreamEvent(event *pbwechaty.EventResponse) {
+  log.Printf("PuppetHostie onGrpcStreamEvent({type:%s payload:%s})", event.Type, event.Payload)
+
+  if event.Type != pbwechaty.EventType_EVENT_TYPE_HEARTBEAT {
```

Comment:
Please check here `!=` ?

Reply:
@huan The `! =` here refers to TypeScript hosite, can you explain it? Why not `heartbeat` but emit` heartbeat`

Reply:
@huan Here you need to confirm, and then I will mark the resolve.

Reply:
Confirm. The logic is right. 

The events other than heartbeat are treated as the heartbeat because all events are coming from the GRPC service, which means the puppet service has sent data through the network to our puppet hostie. So all events are treated as the heartbeat.

We use `!=` is because we need to prevent to emit the `heartbeat` event twice.